cocotb/types/array.py

```diff
@@ -0,0 +1,314 @@
+            ValueError: When the value prevents construction into an element.
+        """
+        return elem
+
+    @staticmethod
+    def _construct_range(rng: Any) -> Range:
+        """
+        Constructs the range portion of the array.
+
+        For overriding by subclasses.
+        Used by the constructor when setting the range
+        and by __setitem__ when creating a slice.
+
+        Args:
+            rng: Any object that the implementation can turn into a :class:`Range` object.
+
+        Returns:
+            The :class:`Range` object describing the array's indexing scheme.
+
+        Raises:
+            TypeError: When type isn't supported.
+            ValueError: When the value prevents construction into a range.
+        """
+        if isinstance(rng, Range):
+            return rng
+        raise TypeError("range argument must be of type 'Range'")
+
+    @cached_property
+    def left(self) -> int:
+        """Leftmost index of the array."""
+        return self._range.left
+
+    @cached_property
+    def direction(self) -> str:
+        """``'to'`` if indexes are ascending, ``'downto'`` otherwise."""
+        return self._range.direction
+
+    @cached_property
+    def right(self) -> int:
+        """Rightmost index of the array."""
+        return self._range.right
+
+    @cached_property
+    def length(self):
+        """Length of the array."""
+        return self._range.length
+
+    @cached_property
+    def range(self) -> Range:
+        """:class:`Range` of the indexes of the array."""
+        return self._range
+
+    def __len__(self) -> int:
+        return self.length
+
+    def __iter__(self) -> Iterator[Any]:
+        return iter(self._value)
+
+    def __reversed__(self) -> Iterator[Any]:
+        return reversed(self._value)
+
+    def __contains__(self, item: Any) -> bool:
+        return item in self._value
+
+    def __eq__(self, other: Any) -> bool:
+        try:
+            return all(a == b for a, b in zip_longest(self, other, fillvalue=object()))
+        except TypeError:
+            return NotImplemented
+
+    @overload
+    def __getitem__(self, item: int) -> Any:
+        pass  # pragma: no cover
+
+    @overload
+    def __getitem__(self, item: slice) -> "Array":
+        pass  # pragma: no cover
+
+    def __getitem__(self, item):
+        if isinstance(item, int):
+            idx = self._translate_index(item)
+            return self._value[idx]
+        elif isinstance(item, slice):
+            start = item.start or self.left
+            stop = item.stop or self.right
+            if item.step is not None:
+                raise IndexError("do not specify step")
+            start_i = self._translate_index(start)
+            stop_i = self._translate_index(stop)
+            if start_i > stop_i:
+                raise IndexError("slice direction does not match array direction")
+            value = self._value[start_i : stop_i + 1]
```

Comment:
I don't see any documentation describing this behavior; at the very least, this needs a big `.. warning` in the docstring saying that unlike _all builtin python objects_, the slice indices are inclusive in the upper bound.

I'm personally dubious whether this is a good idea at all, while convenient for an HDL programmer, it feels like a trap for python programmers, and for passing these objects into functions expecting well-behaved sequences.

One compromise would be to reserve `value.v[x:y]` for verilog-style inclusive indexing, and use `value[x:y]` for python-style index.

Reply:
I added a sentence stating that slicing works just like the `Range` object being fed to it, it uses an inclusive right bound. This parallels how slicing regular Python sequence types and `range` work. The whole point of `Range` was to describe index sequences with inclusive right bounds using typical VHDL parlance (or Verilog-esque automatic directionality); so slicing should behave the same way.

The first thing I tried was just using `range` and non-inclusive right bound. It's very awkward for people accustomed to HDL. Non-inclusive right bounds and non-0-based indexing don't play well together and it isn't very ergonmic. Every non-0-based indexes language I know of uses inclusive bounds. If supporting Python's semantics here is worth doing (I don't agree that it is), it should be the second-class syntax.

Nothing in `collections.abc` or any documentation I see states explicitly that higher-order functions that take `Sequence`s will assume a particular indexing scheme. The default mixins of `Sequence` do though... which is why I had to override them. I guess it's a good question if this is even a `Sequence` or not?

Reply:
> The default mixins of Sequence do though

As does the builtin `bisect` module - so we probably need to warn against using that in the docstring too.

Reply:
Added additional warning information on differences with builtin sequence types.